File: api-js/src/paywall.js
const path = require("path");
const fs = require("fs");
const os = require("os");
const {runNativeblocksCommand} = require("./run-nativeblocks");

const getPaywallFrame = async () => {
  const features = [
    "Secured 2FA storage",
    "End-to-end encryption",
    "Private Browser",
    "No ads and no limits",
  ];

  const baseFrame = {
    $schema: "https://schema-sample.alireza-527.workers.dev/",
    name: "PaywallFrame",
    route: "paywall",
    isStarter: false,
    type: "FRAME",
    variables: [
      {
        key: "root_visible",
        value: "true",
        type: "BOOLEAN",
      },
      {
        key: "main_container_visible",
        value: "true",
        type: "BOOLEAN",
      },
      {
        key: "features_container_visible",
        value: "true",
        type: "BOOLEAN",
      },
      {
        key: "check_icon_source",
        value: "check",
        type: "STRING",
      },
      {
        key: "yearly_price",
        value: `Yearly: 29.99`,
        type: "STRING",
      },
      {
        key: "monthly_price",
        value: `Monthly: 35.99`,
        type: "STRING",
      },
      ...features.map((_, index) => ({
        key: `feature_${index + 1}_visible`,
        value: "true",
        type: "BOOLEAN",
      })),
      ...features.map((feature, index) => ({
        key: `feature_${index + 1}_text`,
        value: feature,
        type: "STRING",
      })),
    ],
    blocks: [
      {
        keyType: "ROOT",
        key: "root",
        visibilityKey: "root_visible",
        slot: "content",
        actions: [],
        data: [],
        properties: [],
        slots: [{slot: "content"}],
        integrationVersion: 1,
        blocks: [
          {
            keyType: "COLUMN",
            key: "main_container",
            visibilityKey: "main_container_visible",
            slot: "content",
            actions: [],
            data: [],
            properties: [],
            slots: [{slot: "content"}],
            integrationVersion: 1,
            blocks: [
              {
                keyType: "COLUMN",
                key: "features_container",
                visibilityKey: "features_container_visible",
                slot: "content",
                actions: [],
                data: [],
                properties: [
                  {
                    key: "width",
                    valueMobile: "match",
                    valueTablet: "match",
                    valueDesktop: "match",
                    type: "STRING",
                  },
                ],
                slots: [
                  {
                    slot: "content",
                  },
                ],
                integrationVersion: 1,
                blocks: features.map((_, index) => ({
                  keyType: "ROW",
                  key: `feature_${index + 1}_row`,
                  visibilityKey: `feature_${index + 1}_visible`,
                  slot: "content",
                  actions: [],
                  data: [],
                  properties: [
                    {
                      key: "width",
                      valueMobile: "match",
                      valueTablet: "match",
                      valueDesktop: "match",
                      type: "STRING",
                    },
                  ],
                  slots: [{slot: "content"}],
                  integrationVersion: 1,
                  blocks: [
                    {
                      keyType: "IMAGE",
                      key: `check_${index + 1}`,
                      visibilityKey: `feature_${index + 1}_visible`,
                      slot: "content",
                      actions: [],
                      data: [
                        {
                          key: "imageUrl",
                          value: "check_icon_source",
                          type: "STRING",
                        },
                      ],
                      properties: [],
                      slots: [],
                      integrationVersion: 1,
                      blocks: [],
                    },
                    {
                      keyType: "TEXT",
                      key: `feature_${index + 1}_text`,
                      visibilityKey: `feature_${index + 1}_visible`,
                      slot: "content",
                      actions: [],
                      data: [
                        {
                          key: "text",
                          value: `feature_${index + 1}_text`,
                          type: "STRING",
                        },
                      ],
                      properties: [
                        {
                          key: "textColor",
                          valueMobile: "#212121",
                          valueTablet: "#212121",
                          valueDesktop: "#212121",
                          type: "STRING",
                        },
                      ],
                      slots: [],
                      integrationVersion: 1,
                      blocks: [],
                    },
                  ],
                })),
              },
              {
                keyType: "ROW",
                key: "pricing_buttons",
                visibilityKey: "main_container_visible",
                slot: "content",
                actions: [],
                data: [],
                properties: [],
                slots: [
                  {
                    slot: "content",
                  },
                ],
                integrationVersion: 1,
                blocks: [
                  {
                    keyType: "BUTTON",
                    key: "yearly_button",
                    visibilityKey: "main_container_visible",
                    slot: "content",
                    actions: [],
                    data: [
                      {
                        key: "text",
                        value: "yearly_price",
                        type: "STRING",
                      },
                      {
                        key: "enable",
                        value: "main_container_visible",
                        type: "BOOLEAN",
                      },
                    ],
                    properties: [],
                    slots: [],
                    integrationVersion: 1,
                    blocks: [],
                  },
                  {
                    keyType: "BUTTON",
                    key: "monthly_button",
                    visibilityKey: "main_container_visible",
                    slot: "content",
                    actions: [],
                    data: [
                      {
                        key: "text",
                        value: "monthly_price",
                        type: "STRING",
                      },
                      {
                        key: "enable",
                        value: "main_container_visible",
                        type: "BOOLEAN",
                      },
                    ],
                    properties: [],
                    slots: [],
                    integrationVersion: 1,
                    blocks: [],
                  },
                ],
              },
            ],
          },
          {
            keyType: "COLUMN",
            key: "additional_container",
            visibilityKey: "main_container_visible",
            slot: "content",
            actions: [],
            data: [],
            properties: [],
            slots: [],
            integrationVersion: 1,
            blocks: [],
          },
        ],
      },
    ],
  };

  const tempFilePath = await createTempFile(baseFrame);
  const response = await runNativeblocksCommand(tempFilePath);
  fs.unlinkSync(tempFilePath);
  return response;
};

async function createTempFile(jsonData) {
  const tempDir = os.tmpdir();
  const tempFilePath = path.join(tempDir, `temp-${Date.now()}.json`);

  await fs.promises.writeFile(tempFilePath, JSON.stringify(jsonData, null, 2));
  return tempFilePath;
}

module.exports = {getPaywallFrame};

File: api-go/paywall.go
package main

import (
	"fmt"
	"github.com/nativeblocks/nativeblocks-example/api-go/integration"
)

type Variable struct {
	Key   string      `json:"key"`
	Value interface{} `json:"value"`
	Type  string      `json:"type"`
}

type Frame struct {
	Schema    string        `json:"$schema"`
	Name      string        `json:"name"`
	Route     string        `json:"route"`
	IsStarter bool          `json:"isStarter"`
	Type      string        `json:"type"`
	Variables []Variable    `json:"variables"`
	Blocks    []interface{} `json:"blocks"`
}

func GetPaywallFrame() (map[string]interface{}, error) {
	features := []string{
		"Secured 2FA storage",
		"End-to-end encryption",
		"Private Browser",
		"No ads and no limits",
	}

	variables := []Variable{
		{Key: "root_visible", Value: "true", Type: "BOOLEAN"},
		{Key: "main_container_visible", Value: "true", Type: "BOOLEAN"},
		{Key: "features_container_visible", Value: "true", Type: "BOOLEAN"},
		{Key: "check_icon_source", Value: "check", Type: "STRING"},
		{Key: "yearly_price", Value: "Yearly: 29.99", Type: "STRING"},
		{Key: "monthly_price", Value: "Monthly: 35.99", Type: "STRING"},
		{Key: "button_visible", Value: "true", Type: "BOOLEAN"},
		{Key: "button_enable", Value: "true", Type: "BOOLEAN"},
	}

	for i := range features {
		variables = append(variables, Variable{
			Key:   fmt.Sprintf("feature_%d_visible", i+1),
			Value: "true",
			Type:  "BOOLEAN",
		})
	}

	for i, feature := range features {
		variables = append(variables, Variable{
			Key:   fmt.Sprintf("feature_%d_text", i+1),
			Value: feature,
			Type:  "STRING",
		})
	}

	root := integration.NewRootBlock("ROOT", "root_visible", "null")
	root.ModifyProperty("width", "match", "match", "match")
	root.ModifyProperty("height", "match", "match", "match")

	button := integration.NewButtonBlock("BUTTON", "button_visible", "content")
	button.ModifyProperty("width", "match", "match", "match")
	button.ModifyProperty("height", "warp", "warp", "warp")
	button.AssignData("text", "yearly_price")
	button.AssignData("enable", "button_enable")

	offerList := integration.NewOfferListAction("offerList", "END")
	offerList.ModifyProperty("endpoint", "http://example.com")
	buttonTriggers := make([]interface{}, 0)
	list := append(buttonTriggers, offerList.Build())
	button.AddAction("onClick", list)

	root.AddBlock(button.Build())

	blocks := make([]interface{}, 0)
	blocks = append(blocks, root.Build())

	frame := &Frame{
		Schema:    "https://raw.githubusercontent.com/nativeblocks/nativeblocks-examples/main/api-ts/.nativeblocks/schema.json",
		Name:      "PaywallFrame",
		Route:     "paywall",
		IsStarter: false,
		Type:      "FRAME",
		Variables: variables,
		Blocks:    blocks,
	}
	return BuildWithNativeblocks(frame)
}
